nexus/src/config.rs

```diff
@@ -150,74 +141,28 @@ pub struct Config {
     pub tunables: Tunables,
 }
 
-#[derive(Debug)]
-pub struct InvalidTunable {
-    tunable: String,
-    message: String,
-}
-
-impl std::fmt::Display for InvalidTunable {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "invalid \"{}\": \"{}\"", self.tunable, self.message)
-    }
-}
-
-impl std::error::Error for InvalidTunable {}
-
-#[derive(Debug)]
-pub struct LoadError {
-    path: PathBuf,
-    kind: LoadErrorKind,
-}
-#[derive(Debug)]
-pub enum LoadErrorKind {
-    Io(std::io::Error),
-    Parse(toml::de::Error),
-    InvalidTunable(InvalidTunable),
-}
-
-impl From<(PathBuf, std::io::Error)> for LoadError {
-    fn from((path, err): (PathBuf, std::io::Error)) -> Self {
-        LoadError { path, kind: LoadErrorKind::Io(err) }
-    }
-}
-
-impl From<(PathBuf, toml::de::Error)> for LoadError {
-    fn from((path, err): (PathBuf, toml::de::Error)) -> Self {
-        LoadError { path, kind: LoadErrorKind::Parse(err) }
-    }
-}
-
-impl std::error::Error for LoadError {}
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct Config {
+    /// Configuration parameters known at compile-time.
+    #[serde(flatten)]
+    pub pkg: PackageConfig,
 
-impl fmt::Display for LoadError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match &self.kind {
-            LoadErrorKind::Io(e) => {
-                write!(f, "read \"{}\": {}", self.path.display(), e)
-            }
-            LoadErrorKind::Parse(e) => {
-                write!(f, "parse \"{}\": {}", self.path.display(), e)
-            }
-            LoadErrorKind::InvalidTunable(inner) => {
-                write!(
-                    f,
-                    "invalid tunable \"{}\": {}",
-                    self.path.display(),
-                    inner,
-                )
-            }
-        }
-    }
+    /// A variety of configuration parameters only known at runtime.
+    pub runtime: RuntimeConfig,
 }
 
-impl std::cmp::PartialEq<std::io::Error> for LoadError {
-    fn eq(&self, other: &std::io::Error) -> bool {
-        if let LoadErrorKind::Io(e) = &self.kind {
-            e.kind() == other.kind()
-        } else {
-            false
-        }
+impl Config {
+    /// Load a `PackageConfig` from the given TOML file
```

Comment:
Nit, but this is the top-level `Config`, not the `PackageConfig`. I guess everything is flattened at serialization, so maybe they're the same?

Reply:
Comment updated